File: docs/s1/en/cpp/tutorials/src/example2_arm_manipulation.cpp
#include <chrono>
#include <cctype>
#include <cmath>
#include <cstdlib>
#include <iostream>
#include <stdexcept>
#include <string>
#include <thread>
#include <unordered_map>
#include <vector>

#include "galbot_robot.hpp"
#include "galbot_motion.hpp"

using namespace galbot::sdk;

void print_motion_status(MotionStatus status)
{
    if (status == MotionStatus::SUCCESS) {
        std::cout << "Execution result: SUCCESS, Execution successful\n";
    } else if (status == MotionStatus::TIMEOUT) {
        std::cout << "Execution result: TIMEOUT, Execution timeout\n";
    } else if (status == MotionStatus::FAULT) {
        std::cout << "Execution result: FAULT, Fault occurred, unable to continue execution\n";
    } else if (status == MotionStatus::INVALID_INPUT) {
        std::cout << "Execution result: INVALID_INPUT, Input parameters do not meet requirements\n";
    } else if (status == MotionStatus::INIT_FAILED) {
        std::cout << "Execution result: INIT_FAILED, Internal communication component creation failed\n";
    } else if (status == MotionStatus::IN_PROGRESS) {
        std::cout << "Execution result: IN_PROGRESS, Moving but not in position\n";
    } else if (status == MotionStatus::STOPPED_UNREACHED) {
        std::cout << "Execution result: STOPPED_UNREACHED, Stopped but did not reach target\n";
    } else if (status == MotionStatus::DATA_FETCH_FAILED) {
        std::cout << "Execution result: DATA_FETCH_FAILED, Data acquisition failed\n";
    } else if (status == MotionStatus::PUBLISH_FAIL) {
        std::cout << "Execution result: PUBLISH_FAIL, Data sending failed\n";
    } else if (status == MotionStatus::COMM_DISCONNECTED) {
        std::cout << "Execution result: COMM_DISCONNECTED, Connection failed\n";
    } else {
        std::cout << "Execution result: status=" << static_cast<int>(status) << "\n";
    }
}

inline std::vector<double> quat_normalize(const std::vector<double>& q)
{
    double norm = std::sqrt(
        q[0]*q[0] + q[1]*q[1] +
        q[2]*q[2] + q[3]*q[3]
    );
    if (norm <= 0.0)
        throw std::runtime_error("Zero-norm quaternion");

    return std::vector<double>{ q[0]/norm, q[1]/norm, q[2]/norm, q[3]/norm };
}
inline std::vector<double> quat_conjugate(const std::vector<double>& q)
{
    return std::vector<double>{ -q[0], -q[1], -q[2], q[3] };
}

inline std::vector<double> quat_multiply(const std::vector<double>& q1, const std::vector<double>& q2)
{
    const double x1 = q1[0];
    const double y1 = q1[1];
    const double z1 = q1[2];
    const double w1 = q1[3];
    const double x2 = q2[0];
    const double y2 = q2[1];
    const double z2 = q2[2];
    const double w2 = q2[3];

    return std::vector<double>{
        w1*x2 + x1*w2 + y1*z2 - z1*y2,
        w1*y2 - x1*z2 + y1*w2 + z1*x2,
        w1*z2 + x1*y2 - y1*x2 + z1*w2,
        w1*w2 - x1*x2 - y1*y2 - z1*z2
    };
}

template <typename T>
inline T clamp_val(const T& v, const T& lo, const T& hi)
{
    return (v < lo) ? lo : (v > hi) ? hi : v;
}

inline double orientation_error_angle(const std::vector<double>& A, const std::vector<double>& B)
{
    std::vector<double> qA = quat_normalize({ A[3], A[4], A[5], A[6] });
    std::vector<double> qB = quat_normalize({ B[3], B[4], B[5], B[6] });

    std::vector<double> q_err = quat_multiply(qB, quat_conjugate(qA));
    q_err = quat_normalize(q_err);

    double qw = clamp_val(q_err[3], -1.0, 1.0);

    return 2.0 * std::acos(qw);
}

struct PoseError
{
    double position_error_norm;
    double orientation_error_rad;
    double orientation_error_deg;
};

inline PoseError calculate_error(const std::vector<double>& pose1, const std::vector<double>& pose2)
{
    double dx = pose1[0] - pose2[0];
    double dy = pose1[1] - pose2[1];
    double dz = pose1[2] - pose2[2];

    double pos_err = std::sqrt(dx*dx + dy*dy + dz*dz);
    double rot_err = orientation_error_angle(pose1, pose2);

    return PoseError{
        pos_err,
        rot_err,
        rot_err * 180.0 / M_PI
    };
}

void print_pose_error(const char* label, const PoseError& e)
{
    std::cout << label
              << " position_error_norm: " << e.position_error_norm
              << " m, orientation_error_rad: " << e.orientation_error_rad
              << ", orientation_error_deg: " << e.orientation_error_deg << "\n";
}

void check_robot_safety()
{
    std::cout << "⚠️  Note: 1. Please ensure the robot's emergency stop button is released; 2. Please ensure there are no obstacles in front, back, left, and right of the robot to avoid unexpected situations.\n\n";

    char key;
    for (;;) {
        std::cout << "Please confirm that the robot's emergency stop button is released and there are no obstacles. Continue? (y/n)...";
        std::cin >> key;

        if (std::tolower(static_cast<unsigned char>(key)) == 'y') {
            std::cout << "User confirmed, continuing execution...\n\n";
            break;
        }
        if (std::tolower(static_cast<unsigned char>(key)) == 'n') {
            std::cout << "User not confirmed, program exiting...\n\n";
            std::exit(1);
        }
        std::cout << "Input error, please enter 'y' or 'n'\n\n";
    }
}

void print_pose_line(const char* prefix, const std::vector<double>& pose)
{
    std::cout << prefix;
    for (double v : pose) {
        std::cout << v << " ";
    }
    std::cout << "\n";
}

int main()
{
    check_robot_safety();
    try {
        auto& motion = GalbotMotion::get_instance(MachineType::S1);
        auto& robot = GalbotRobot::get_instance(MachineType::S1);

        struct SdkCleanup {
            GalbotRobot& r;
            explicit SdkCleanup(GalbotRobot& ref) : r(ref) {}
            ~SdkCleanup()
            {
                r.request_shutdown();
                r.wait_for_shutdown();
                r.destroy();
            }
        } cleanup{robot};

        if (motion.init()) {
            std::cout << "GalbotMotion initialization successful\n";
        } else {
            std::cout << "GalbotMotion initialization failed\n";
        }
        if (robot.init()) {
            std::cout << "GalbotRobot initialization successful\n";
        } else {
            std::cout << "GalbotRobot initialization failed\n";
        }

        std::this_thread::sleep_for(std::chrono::milliseconds(3000));

        MotionStatus zero_status = motion.move_whole_body_joint_zero(
            true, 0.2, 15.0, std::make_shared<Parameter>());
        print_motion_status(zero_status);
        if (zero_status != MotionStatus::SUCCESS) {
            std::cout << "❌ move_whole_body_joint_zero failed, stop demo to avoid risky posture.\n";
            return 0;
        }
        std::cout << "✅ whole-body zero completed, start arm manipulation demo from safe pose.\n";

        const std::string target_frame = "EndEffector";
        const std::string reference_frame = "base_link";
        const std::string target_chain = "left_arm";
        const std::string end_link = "left_arm_end_effector_mount_link";

        std::vector<double> original_pose;
        {
            auto ret_pose = motion.get_end_effector_pose_on_chain(
                target_chain, target_frame, reference_frame);
            if (std::get<0>(ret_pose) != MotionStatus::SUCCESS) {
                std::cerr << "❌ Failed to get end-effector pose\n";
                return 0;
            }
            original_pose = std::get<1>(ret_pose);
            std::cout << "✅ Current " << target_chain << " end-effector pose: ";
            print_pose_line("", original_pose);
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(800));

        std::vector<double> target_pose = {
            original_pose[0] + 0.04,
            original_pose[1],
            original_pose[2] + 0.02,
            original_pose[3],
            original_pose[4],
            original_pose[5],
            original_pose[6]
        };
        std::cout << "✅ S1 target pose (relative to zero pose): ";
        print_pose_line("", target_pose);

        std::unordered_map<std::string, std::vector<double>> joint_angles_ik;
        {
            const bool enable_collision_check = false;
            auto ret_ik = motion.inverse_kinematics(
                target_pose,
                {target_chain},
                target_frame,
                reference_frame,
                {},
                enable_collision_check,
                std::make_shared<Parameter>());
            if (std::get<0>(ret_ik) != MotionStatus::SUCCESS) {
                std::cerr << "❌ IK solving failed\n";
                return 0;
            }
            joint_angles_ik = std::get<1>(ret_ik);
            std::cout << "✅ Target " << target_chain << " IK solving successful joint_angles_ik: ";
            print_pose_line("", joint_angles_ik[target_chain]);
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(1000));

        ControlStatus set_joints_status = robot.set_joint_positions(
            joint_angles_ik[target_chain],
            {target_chain},
            {},
            true,
            0.1,
            20.0);
        if (set_joints_status != ControlStatus::SUCCESS) {
            std::cerr << "❌ Setting joint group angles failed\n";
            return 0;
        }
        std::cout << "✅ Setting " << target_chain << " joint group angles successful.\n";
        std::this_thread::sleep_for(std::chrono::milliseconds(1000));

        {
            auto ret_verify = motion.get_end_effector_pose_on_chain(
                target_chain, target_frame, reference_frame);
            if (std::get<0>(ret_verify) != MotionStatus::SUCCESS) {
                std::cerr << "❌ Failed to get end-effector pose after set joints\n";
                return 0;
            }
            const std::vector<double>& tgt_pose_ik = std::get<1>(ret_verify);
            std::cout << "✅ Getting " << target_chain << " end-effector pose successful: ";
            print_pose_line("", tgt_pose_ik);
            PoseError err = calculate_error(tgt_pose_ik, target_pose);
            print_pose_error("End-effector pose error:", err);
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(1000));

        {
            const bool enable_collision_check = false;
            auto ret_fk = motion.forward_kinematics(
                end_link,
                reference_frame,
                joint_angles_ik,
                std::make_shared<Parameter>());
            if (std::get<0>(ret_fk) != MotionStatus::SUCCESS) {
                std::cerr << "❌ FK solving failed\n";
                return 0;
            }
            const std::vector<double>& tgt_pose_fk = std::get<1>(ret_fk);
            std::cout << "✅ Target " << target_chain << " FK solving successful: ";
            print_pose_line("", tgt_pose_fk);
            PoseError err = calculate_error(tgt_pose_fk, target_pose);
            print_pose_error("FK solving error:", err);
        }

        std::this_thread::sleep_for(std::chrono::milliseconds(3000));
        std::cout << "\n";

        {
            const bool enable_collision_check = false;
            MotionStatus ret_restore = motion.set_end_effector_pose(
                original_pose,
                target_chain,
                reference_frame,
                nullptr,
                enable_collision_check,
                true,
                5.0,
                std::make_shared<Parameter>());
            if (ret_restore != MotionStatus::SUCCESS) {
                std::cerr << "❌ Setting end-effector pose failed\n";
                return 0;
            }
            std::cout << "✅ Setting end-effector pose successful: status=" << static_cast<int>(ret_restore) << "\n";
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(1000));

        {
            auto ret_final = motion.get_end_effector_pose_on_chain(
                target_chain, target_frame, reference_frame);
            if (std::get<0>(ret_final) != MotionStatus::SUCCESS) {
                std::cerr << "❌ Failed to get end-effector pose after restore\n";
                return 0;
            }
            const std::vector<double>& original_pose_rec = std::get<1>(ret_final);
            std::cout << "✅ Getting " << target_chain << " end-effector pose successful: ";
            print_pose_line("", original_pose_rec);
            PoseError err = calculate_error(original_pose_rec, original_pose);
            print_pose_error("Restore end-effector pose error:", err);
        }
    } catch (const std::exception& e) {
        std::cerr << "❌ Main program exception: " << e.what() << "\n";
        return 1;
    }

    return 0;
}

File: docs/s1/en/python/tutorials/example2_arm_manipulation.py
"""
Note: When running this example, please ensure the robot's motion control service `/data/galbot/bin/service_motion_plan`,
    robot state publishing service `/data/galbot/bin/robot_state_publish`,
    and hand-eye calibration publishing service `/data/galbot/bin/eyehand_calib_publish` are loaded;
"""
try:
    import galbot_sdk.s1 as gm
    from galbot_sdk.s1 import GalbotMotion
    from galbot_sdk.s1 import GalbotRobot
    from galbot_sdk.s1 import ControlStatus
except ImportError:
    print("Failed to import galbot_sdk, please install it first or check if it is in PYTHONPATH")
    exit(1)

import os

try:
    import numpy as np
except ImportError:
    os.system("pip install numpy")
    import numpy as np

import time
from typing import Sequence, Dict

def printStatus(status):
    if(status == gm.MotionStatus.SUCCESS):
        print("Execution result: SUCCESS, Execution successful")
    elif(status == gm.MotionStatus.TIMEOUT):
        print("Execution result: TIMEOUT, Execution timeout")
    elif(status == gm.MotionStatus.FAULT):
        print("Execution result: FAULT, Fault occurred, unable to continue execution")
    elif(status == gm.MotionStatus.INVALID_INPUT):
        print("Execution result: INVALID_INPUT, Input parameters do not meet requirements")
    elif(status == gm.MotionStatus.INIT_FAILED):
        print("Execution result: INIT_FAILED, Internal communication component creation failed")
    elif(status == gm.MotionStatus.IN_PROGRESS):
        print("Execution result: IN_PROGRESS, Moving but not in position")
    elif(status == gm.MotionStatus.STOPPED_UNREACHED):
        print("Execution result: STOPPED_UNREACHED, Stopped but did not reach target")
    elif(status == gm.MotionStatus.DATA_FETCH_FAILED):
        print("Execution result: DATA_FETCH_FAILED, Data acquisition failed")
    elif(status == gm.MotionStatus.PUBLISH_FAIL):
        print("Execution result: PUBLISH_FAIL, Data sending failed")
    elif(status == gm.MotionStatus.COMM_DISCONNECTED):
        print("Execution result: COMM_DISCONNECTED, Connection failed")

def quat_normalize(q: np.ndarray) -> np.ndarray:
    q = np.array(q, dtype=np.float64)
    return q / np.linalg.norm(q)

def quat_conjugate(q: np.ndarray) -> np.ndarray:
    """
    Calculate the conjugate of a quaternion

    Parameters:
        q (np.ndarray): Input quaternion [x, y, z, w]

    Returns:
        np.ndarray: Conjugate of the quaternion [x, y, z, w]
    """
    qx, qy, qz, qw = q
    return np.array([-qx, -qy, -qz, qw])

def quat_multiply(q1: np.ndarray, q2: np.ndarray) -> np.ndarray:
    """
    Calculate the product of two quaternions

    Parameters:
        q1 (np.ndarray): First quaternion [x, y, z, w]
        q2 (np.ndarray): Second quaternion [x, y, z, w]

    Returns:
        np.ndarray: Product of the two quaternions [x, y, z, w]
    """
    x1, y1, z1, w1 = q1
    x2, y2, z2, w2 = q2
    return np.array([
        w1*x2 + x1*w2 + y1*z2 - z1*y2,
        w1*y2 - x1*z2 + y1*w2 + z1*x2,
        w1*z2 + x1*y2 - y1*x2 + z1*w2,
        w1*w2 - x1*x2 - y1*y2 - z1*z2
    ])

def orientation_error_angle(A: np.ndarray, B: np.ndarray) -> float:
    """
    Calculate the rotation angle error between two quaternions (radians)

    Parameters:
        A (np.ndarray): First quaternion [x, y, z, w]
        B (np.ndarray): Second quaternion [x, y, z, w]

    Returns:
        float: Rotation angle error (radians)
    """
    qA = quat_normalize(A[3:7])
    qB = quat_normalize(B[3:7])

    q_err = quat_multiply(qB, quat_conjugate(qA))
    q_err = quat_normalize(q_err)

    # Numerical stability
    qw = np.clip(q_err[3], -1.0, 1.0)

    angle = 2 * np.arccos(qw)
    return angle  # Unit: radians


def calculate_error(pose1: Sequence[float], pose2: Sequence[float]) -> Dict[str, float]:
    """
    Calculate position error and rotation error between two poses (radians)

    Parameters:
        pose1 (Sequence[float]): First pose [x, y, z, qx, qy, qz, qw]
        pose2 (Sequence[float]): Second pose [x, y, z, qx, qy, qz, qw]

    Returns:
        dict: Dictionary containing position error (meters) and rotation error (radians)
    """
    A, B = np.array(pose1), np.array(pose2)
    pos_err = np.linalg.norm(A[:3] - B[:3])
    rot_err = orientation_error_angle(A, B)

    return {
        "position_error_norm": pos_err,
        "orientation_error_rad": rot_err,
        "orientation_error_deg": np.degrees(rot_err)
    }

def check_robot_safety():
    """Check if the robot is safe"""
    # Prompt for precautions
    print("⚠️  Note: 1. Please ensure the robot's emergency stop button is released; 2. Please ensure there are no obstacles in front, back, left, and right of the robot to avoid unexpected situations.")
    while True:
        key = input("Please confirm that the robot's emergency stop button is released and there are no obstacles. Continue? (y/n)...")
        if key == 'y':
            print("User confirmed, continuing execution...")
            break
        elif key == 'n':
            print("User not confirmed, program exiting...")
            exit(1)
        else:
            print("Input error, please enter 'y' or 'n'")

def main():
    check_robot_safety()
    try:
        # Get GalbotMotion singleton and initialize
        motion = GalbotMotion()
        robot = GalbotRobot()

        if motion.init():
            print("GalbotMotion initialization successful")
        else:
            print("GalbotMotion initialization failed")
        if robot.init():
            print("GalbotRobot initialization successful")
        else:
            print("GalbotRobot initialization failed")

        # Program starts immediately, wait for data readiness time
        time.sleep(3)

        # Move whole body to predefined zero(home) configuration before arm demo.
        status = motion.move_whole_body_joint_zero(
            is_blocking=True,
            leg_head_speed_rad_s=0.2,
            leg_head_timeout_s=15.0,
            params=gm.Parameter()
        )
        printStatus(status)
        if status != gm.MotionStatus.SUCCESS:
            print("❌ move_whole_body_joint_zero failed, stop demo to avoid risky posture.")
            return
        print("✅ whole-body zero completed, start arm manipulation demo from safe pose.")

        # Define target chain name, target pose, reference pose, end link
        target_frame = "EndEffector"
        reference_frame = "base_link"
        target_chain = "left_arm"
        end_link = "left_arm_end_effector_mount_link"

        # 1. Get current left_arm end-effector pose
        try:
            status, original_pose = motion.get_end_effector_pose_on_chain(
                chain_name=target_chain,
                frame_id=target_frame,
                reference_frame=reference_frame
            )
            assert status == gm.MotionStatus.SUCCESS, "Failed to get end-effector pose"
            print(f"✅ Current {target_chain} end-effector pose: {original_pose}")
            time.sleep(0.8)
        except Exception as e:
            print(f"❌ Exception getting end-effector pose by chain name: {e}")
            return

        # Build S1 demo target pose from zero-start current pose instead of hard-coded absolute pose.
        # This avoids using model-specific (e.g. G1-like) constants in S1 tutorial.
        chain_pose_baselink = {
            "left_arm": [
                original_pose[0] + 0.04,  # x forward 4 cm
                original_pose[1],         # keep lateral position
                original_pose[2] + 0.02,  # z up 2 cm
                original_pose[3],
                original_pose[4],
                original_pose[5],
                original_pose[6]
            ]
        }
        print(f"✅ S1 target pose (relative to zero pose): {chain_pose_baselink[target_chain]}")

        # 2. Solve joint angles based on target pose IK and verify the solution
        # 2.1 Solve joint angles joint_angles_ik for target pose through IK
        try:
            status, joint_angles_ik = motion.inverse_kinematics(
                target_pose=chain_pose_baselink[target_chain],
                chain_names=[target_chain],
                target_frame=target_frame,
                reference_frame=reference_frame,
                enable_collision_check=False # Disable collision detection
            )
            assert status == gm.MotionStatus.SUCCESS, "IK solving failed"
            print(f"✅ Target {target_chain} IK solving successful joint_angles_ik: {joint_angles_ik}")
            time.sleep(1)
        except Exception as e:
            print(f"❌ IK solving exception: {e}")

        # 2.2 Set end-effector pose to target pose tgt_pose_ik by setting joint group angles joint_angles_ik
        try:
            status = robot.set_joint_positions(
                joint_angles_ik[target_chain], 
                [target_chain], 
                [], 
                True,
                0.1,
                20.0,
            )
            assert status == ControlStatus.SUCCESS, "Setting joint group angles failed"
            print(f"✅ Setting {target_chain} joint group angles successful.")
            time.sleep(1)
        except Exception as e:
            print(f"❌ Setting {target_chain} joint group angles exception: {e}")

        # 2.3 Verify whether the set joint group angles are consistent with the solved angles
        try:
            status, tgt_pose_ik = motion.get_end_effector_pose_on_chain(
                chain_name=target_chain,
                frame_id=target_frame,
                reference_frame=reference_frame
            )
            assert status == gm.MotionStatus.SUCCESS, "Failed to get end-effector pose"
            print(f"✅ Getting {target_chain} end-effector pose successful: {tgt_pose_ik}")
            time.sleep(1)

            error = calculate_error(tgt_pose_ik, chain_pose_baselink[target_chain])
            print(f"End-effector pose error: {error}")
        except Exception as e:
            print(f"❌ Getting {target_chain} end-effector pose exception: {e}")

        # 2.4 Verify whether the end-effector pose tgt_pose_fk corresponding to joint group angles joint_angles_ik solved by FK is consistent with target pose tgt_pose_ik
        try:
            status, tgt_pose_fk = motion.forward_kinematics(
                target_frame=end_link,
                reference_frame=reference_frame,
                joint_state=joint_angles_ik,
                params=gm.Parameter()
            )
            assert status == gm.MotionStatus.SUCCESS, "FK solving failed"
            print(f"✅ Target {target_chain} FK solving successful: {tgt_pose_fk}")
            time.sleep(1)

            error = calculate_error(tgt_pose_fk, chain_pose_baselink[target_chain])
            print(f"FK solving error: {error}")
        except Exception as e:
            print(f"❌ FK solving exception: {e}")

        time.sleep(3)
        print()

        # 3. Restore to original pose by setting end-effector pose
        # 3.1 Set end-effector pose to restore to original pose
        try:
            status = motion.set_end_effector_pose(
                target_pose=original_pose,
                end_effector_frame=target_chain,
                reference_frame=reference_frame,
                enable_collision_check=False,
                is_blocking=True,
                timeout=5.0,
                params=gm.Parameter()
            )
            assert status == gm.MotionStatus.SUCCESS, "Setting end-effector pose failed"
            print(f"✅ Setting end-effector pose successful: status={status}")
            time.sleep(1)
        except Exception as e:
            print(f"❌ Setting {target_chain} end-effector pose exception: {e}")

        # 3.2 Get end-effector pose and verify whether it has been restored to original pose
        try:
            status, original_pose_rec = motion.get_end_effector_pose_on_chain(
                chain_name=target_chain,
                frame_id=target_frame,
                reference_frame=reference_frame
            )
            assert status == gm.MotionStatus.SUCCESS, "Failed to get end-effector pose"
            print(f"✅ Getting {target_chain} end-effector pose successful: {original_pose_rec}")
            time.sleep(0.8)
            
            error = calculate_error(original_pose_rec, original_pose)
            print(f"Restore end-effector pose error: {error}")
        except Exception as e:
            print(f"❌ Setting end-effector pose exception: {e}")
    
    except Exception as e:
        print(f"❌ Main program exception: {e}")
    finally:
        robot.request_shutdown()
        robot.wait_for_shutdown()
        robot.destroy()

if __name__=="__main__":
    main()